ASP-based solver: always consider version of installed packages

alalazo opened the pull request:
fixes #29201
fixes #30064

Explicitly add facts for versions of installed software when using the `--reuse` option, so that we could consider versions that are not declared in `package.py`

Modifications:
- [x] Fixed the bug
- [x] Add unit test

alalazo committed:
ASP-based solver: always consider version of installed packages

fixes #29201

Explicitly add facts for versions of installed software when
using the --reuse option, so that we could consider versions
that are not declared in package.py

alalazo committed:
Add unit test

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1654,6 +1656,12 @@ def _facts_from_concrete_spec(self, spec, possible):
 
             # add OS to possible OS's
             for dep in spec.traverse():
+                self.possible_versions[dep.name].add(dep.version)
```
The comment above this block probably should change to reflect the new content, maybe

```
# ensure all attributes of installed specs are considered possible values.

alalazo replied:
Fixed in 7727cd1

becker33 reviewed (commented):
Have you tested the following scenario:

Package foo declares versions a, b
foo@c installed
foo@b preferred
foo@b conflicts with something in spec

I think in that case this will lead to concretizing `foo@c` even though it's only a valid version for reusing.

alalazo committed:
Address review comments

alalazo commented:
> Have you tested the following scenario: [ ... ]

That scenario was working, but I came up with something along the same lines that wasn't. Fixed and added a new unit test for it. Thanks!

becker33 commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1485,3 +1525,43 @@ def test_conditional_values_in_conditional_variant(self):
+    def test_installed_version_is_selected_only_for_reuse(
+            self, mutable_database, repo_with_changing_recipe
+    ):
+        """Test that we can reuse installed specs with versions not
+        declared in package.py
+        """
```
This docstring looks like it was mistakenly copied from the previous test.

```suggestion
        """Test that we cannot build with versions only declared for reuse.
        """
```

alalazo replied:
Fixed in 9876f66

alalazo committed:
Fix docstring
